rust/flexbuffers/src/builder/map.rs

```diff
@@ -0,0 +1,122 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::{Builder, Pushable, Value, VectorBuilder};
+
+/// Builds a Flexbuffer map. When this is dropped, or `end` is called, the map is
+/// commited to the buffer. If this map is the root of the flexbuffer then the
+/// root is written and the flexbuffer is complete.
+/// WARNING: Duplicate keys results in a panic.
+pub struct MapBuilder<'a> {
+    pub(super) builder: &'a mut Builder,
+    // If the root is this map then start == None. Otherwise start is the
+    // number of values in the 'values stack' before adding this map.
+    pub(super) start: Option<usize>,
+}
+impl<'a> MapBuilder<'a> {
+    /// Push `p` onto this map with key `key`.
+    /// This will panic if `key` contains internal nulls.
+    pub fn push<P: Pushable>(&mut self, key: &str, p: P) {
+        self.builder.push_key(key);
+        self.builder.push(p);
+    }
+    /// Starts a nested vector which will be pushed onto this map
+    /// with key `key`.
+    /// This will panic if `key` contains internal nulls.
+    pub fn nest_vector(&mut self, key: &str) -> VectorBuilder {
+        // Push the key that refers to this nested vector.
+        self.builder.push_key(key);
+        // Nested vector.
+        let start = Some(self.builder.values.len());
+        VectorBuilder {
+            builder: &mut self.builder,
+            start,
+        }
+    }
+    /// Starts a nested map which which will be pushed onto this map
+    /// with key `key`.
+    /// This will panic if `key` contains internal nulls.
+    pub fn nest_map(&mut self, key: &str) -> MapBuilder {
+        // Push the key that refers to this nested vector.
+        self.builder.push_key(key);
+        // Nested map.
+        let start = Some(self.builder.values.len());
+        MapBuilder {
+            builder: &mut self.builder,
+            start,
+        }
+    }
+    /// `end` sorts the map by key and writes it to the buffer. This happens anyway
+    /// when the map builder is dropped.
+    pub fn end(self) {}
+}
+impl<'a> Drop for MapBuilder<'a> {
+    fn drop(&mut self) {
+        self.builder.end_map_or_vector(true, self.start);
+    }
+}
+
+// Read a known key or string inside the buffer.
+pub(super) unsafe fn read_trusted_str(buffer: &[u8], idx: usize, len: usize) -> &str {
```

Comment:
I think I'd rather keep unsafe out of this if possible. Perhaps by using std::str::from_utf8 and panic'ing on an error. WDYT?

Reply:
I'm doing the unsafe version because the main slowdown, when I profiled bin/monster_profile.rs, was strlen and validate_utf8. We can trust the strings in the buffer to be valid rust str since we wrote it there. That said, comparing [u8] instead of &str will get rid of the use of unsafe. Its just less explicit.